python/cuml/preprocessing/text/stem/porter_stemmer.py

```diff
@@ -0,0 +1,801 @@
+
+        stem = replace_suffix(word_str_ser, "eed", "", valid_mask)
+        measure_mask = measure_gt_n(stem, 0)
+
+        valid_mask = measure_mask & suffix_mask & can_replace_mask
+        # adding ee series to stem
+        word_str_ser = replace_suffix(word_str_ser, "eed", "ee", valid_mask)
+
+        # to be consistent with nltk we dont replace
+        # if word.endswith('eed') we stop proceesing
+        can_replace_mask = can_replace_mask & cudf.logical_not(suffix_mask)
+
+        # rule 2
+        #    (*v*) ED  ->   plastered ->  plaster
+        #                   bled      ->  bled
+
+        ed_suffix_mask = ends_with_suffix(word_str_ser, "ed")
+        intermediate_stem = replace_suffix(
+            word_str_ser, "ed", "", ed_suffix_mask & can_replace_mask
+        )
+        vowel_mask = contains_vowel(intermediate_stem)
+
+        rule_2_mask = vowel_mask & ed_suffix_mask & can_replace_mask
+
+        # rule 3
+
+        #    (*v*) ING ->  motoring  ->  motor
+        #                   sing      ->  sing
+        ing_suffix_mask = ends_with_suffix(word_str_ser, "ing")
+        intermediate_stem = replace_suffix(
+            word_str_ser, "ing", "", ing_suffix_mask & can_replace_mask
+        )
+        vowel_mask = contains_vowel(intermediate_stem)
+        rule_3_mask = vowel_mask & ing_suffix_mask & can_replace_mask
+
+        rule_2_or_rule_3_mask = rule_2_mask | rule_3_mask
+
+        # replace masks only if rule_2_or_rule_3_mask
+        intermediate_stem_1 = replace_suffix(
+            word_str_ser, "ed", "", rule_2_mask
+        )
+        intermediate_stem_2 = replace_suffix(
+            intermediate_stem_1, "ing", "", rule_3_mask
+        )
+
+        can_replace_mask = can_replace_mask & rule_2_or_rule_3_mask
+        return apply_rule_list(
+            intermediate_stem_2,
+            [
+                ("at", "ate", None),  # AT -> ATE
+                ("bl", "ble", None),  # BL -> BLE
+                ("iz", "ize", None),  # IZ -> IZE
+                # (*d and not (*L or *S or *Z))
+                # -> single letter
+                (
+                    "*d",
+                    -1,  # intermediate_stem[-1],
+                    lambda stem: last_char_not_in(
+                        stem, characters=["l", "s", "z"]
+                    ),
+                ),
+                # (m=1 and *o) -> E
+                (
+                    "",
+                    "e",
+                    lambda stem: measure_eq_n(stem, n=1) & ends_cvc(stem),
+                ),
+            ],
+            can_replace_mask,
+        )[0]
+
+    def _step1c(self, word_str_ser, can_replace_mask=None):
+        """Implements Step 1c from "An algorithm for suffix stripping"
+
+        From the paper:
+
+        Step 1c
+
+            (*v*) Y -> I                    happy        ->  happi
+                                            sky          ->  sky
+        """
+        if can_replace_mask is None:
```

Comment:
These two lines are used quite a bit. It might be helpful to encapsulate them behind a function in case it ever needs to be updated. Maybe something like `can_replace_mask = build_can_replace_mask(word_str_ser, can_replace_mask)`.

Reply:
Resolved by https://github.com/rapidsai/cuml/pull/2476/commits/2c8edc2b030424df613ea237f321a885704b1863